crates/app/src/builder.rs

```diff
@@ -1,34 +1,62 @@
 // SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
 // SPDX-License-Identifier: AGPL-3.0-only
 
-use super::{handle, RepoStore, TagStore, TreeStore};
+use super::{handle, RepoStoreKey, TagStoreKey, TreeStoreKey};
 
-use std::sync::Arc;
+use drawbridge_store::{New, Storage};
+
+use std::error::Error;
+use std::ffi::OsStr;
+use std::fs::{create_dir, File};
+use std::path::Path;
 
 use axum::handler::Handler;
 use axum::routing::IntoMakeService;
 use axum::{Extension, Router};
+use cap_async_std::fs::Dir;
+
+pub struct Builder<'a> {
+    data_directory: &'a Path,
+}
 
-#[derive(Default)]
-pub struct Builder;
+impl<'a> Default for Builder<'a> {
+    fn default() -> Self {
+        Self {
+            data_directory: Path::new("drawbridge-data"),
+        }
+    }
+}
 
-impl Builder {
+impl<'a> Builder<'a> {
+    /// Constructs a new [Builder].
     pub fn new() -> Self {
-        Self
+        Default::default()
     }
 
-    // TODO: Add configuration functionality
+    /// The directory to be used for all data storage and configuration.
+    pub fn data_directory<S: AsRef<OsStr>>(mut self, path: &'a S) -> Self {
+        self.data_directory = Path::new(path);
+        self
+    }
 
     /// Builds the application and returns Drawbridge instance as a [tower::MakeService].
-    pub fn build(self) -> IntoMakeService<Router> {
-        let repos: Arc<RepoStore> = Default::default();
-        let tags: Arc<TagStore> = Default::default();
-        let trees: Arc<TreeStore> = Default::default();
-        Router::new()
+    pub fn build(self) -> Result<IntoMakeService<Router>, Box<dyn Error>> {
+        let path = Path::new(&self.data_directory);
+
+        if !path.exists() {
+            create_dir(path)?;
+        }
+
+        let root = Dir::from_std_file(File::open(path)?.into());
+        let repo_store = Storage::<RepoStoreKey>::new_arc(root.clone(), "repo".to_string());
+        let tag_store = Storage::<TagStoreKey>::new_arc(root.clone(), "tag".to_string());
+        let tree_store = Storage::<TreeStoreKey>::new_arc(root, "tree".to_string());
```

Comment:
If you'd prefer that these stay as typedefs rather than distinct types, then I think I'd prefer the previous approach, in conjunction with getting rid of the `New` trait and just defining `::new` in inherent impls on Filesystem and Memory. Still do the `&str` thing mentioned above, though.

Reply:
I also feel like the New trait is kinda weird but doesn't it help keep Filesystem and Memory in sync? We probably would want to create them in the same way right?

Reply:
I don't think `New` is particularly helping to keep anything in sync. The signature of `new` is trivial and it would be quite obvious if anything got out of sync there, as long as we're ever compiling the memory feature in CI (which we should be).